### ISSUE TITLE
Implement QueueBackgroundWorkItem

### ISSUE THREAD

**Issue by davidfowl** [aspnet/Hosting#1280]
Gives user code the ability to schedule a background task that's aware of shutdown semantics. Here's the original blog post of that feature in System.Web.

https://blogs.msdn.microsoft.com/webdev/2014/06/04/queuebackgroundworkitem-to-reliably-schedule-and-run-background-processes-in-asp-net/

@Tratcher wrote a prototype of it here https://github.com/aspnet/Docs/issues/3352#issuecomment-333593326. 

A few initial thoughts on the version that should exist here:
- It shouldn't be tied to anything Web specific, this is a generic host feature.
- We should think about the fact that people will want to use services in these background operations. That might mean that we need to provide a way to active a "Task" so it's not just a func. e.g. QueueBackgroundWorkItem<WorkItem>(arg1, arg2), etc.

**Comment by muratg** [comment]
Tentatively putting this in 2.2.

**Comment by NickCraver** [comment]
Can we please consider this in 2.2? We're working on migrating now, and this is the primary way some of our polling engines for monitoring applications work with System.Web today. If this isn't landing, what is the recommended alternative? Is there one in-box?

**Comment by Tratcher** [comment]
Have you tried the prototype? https://github.com/aspnet/Docs/issues/3352#issuecomment-333593326

**Comment by NickCraver** [comment]
@Tratcher I haven't yet, thanks! I'll see about moving one of our major apps onto ASP.NET Core and giving it ago soon as time allows me to complete it and post feedback here. It's used *a lot* so should give us some good data there. I'll report back!

**Comment by muratg** [comment]
@DamianEdwards @glennc for their thoughts.

**Comment by vegardlarsen** [comment]
I have a suggestion as to how to get services properly injected into the background tasks. My suggestion adds a bit of implementation complexity, but still allows the same easy usage as the current prototype does. It is based directly off of the work @Tratcher did, and caused directly by needing to get proper dependencies into the background worker (as @davidfowl mentioned).

The core of my suggestion is to split the background task into two separate parts: the work order (the request, which are basically the parameters) and the worker (which gets dependency injected). The work order can be serializable, which means it can be put into an external queuing service (this does not apply for the task example).

```cs
public interface IBackgroundWorkOrder { }

public interface IBackgroundWorkOrder<TWorkOrder, TWorker> : IBackgroundWorkOrder
    where TWorker : IBackgroundWorker<TWorkOrder, TWorker>
    where TWorkOrder : IBackgroundWorkOrder<TWorkOrder, TWorker>
{
}

public interface IBackgroundWorker { }

public interface IBackgroundWorker<TWorkOrder, TWorker> : IBackgroundWorker
    where TWorker : IBackgroundWorker<TWorkOrder, TWorker>
    where TWorkOrder : IBackgroundWorkOrder<TWorkOrder, TWorker>
{
    Task DoWork(TWorkOrder order, CancellationToken cancellationToken);
}

public interface IBackgroundTaskQueue
{
    void Queue<TWorkOrder, TWorker>(IBackgroundWorkOrder<TWorkOrder, TWorker> order)
        where TWorker : IBackgroundWorker<TWorkOrder, TWorker>
        where TWorkOrder : IBackgroundWorkOrder<TWorkOrder, TWorker>;

    Task<IBackgroundWorkOrder> DequeueAsync(CancellationToken cancellationToken);
}
```

These interfaces:

- Does go a bit crazy on the generics, but in all instances I've come up with, the generics are automatically resolved when following the basic pattern.
- Lets us pull out the worker `Type` from the `WorkOrder` instance using reflection.
- Lets the `Worker` get a properly typed `WorkOrder` when doing the work.
- `DequeueAsync` returns the marker interface IBackgroundWorkOrder, and can use reflection to find out which worker and work order to use.

Using these interfaces we can implement the same functionality for async functions:

```cs
public static class BackgroundWorkItem
{
    public static void QueueBackgroundWorkItem(
        this IBackgroundTaskQueue queue,
        Func<CancellationToken, Task> method)
    {
        queue.Queue(new WorkOrder(method));
    }

    public class WorkOrder : IBackgroundWorkOrder<WorkOrder, Worker>
    {
        public WorkOrder(Func<CancellationToken, Task> method)
        {
            this.Method = method;
        }

        public Func<CancellationToken, Task> Method { get; }
    }

    public class Worker : IBackgroundWorker<WorkOrder, Worker>
    {
        public async Task DoWork(WorkOrder order, CancellationToken cancellationToken)
        {
            await order.Method.Invoke(cancellationToken);
        }
    }
}
```

The extension method `QueueBackgroundWorkItem` matches the signature provided earlier, so usage should be identical to before.

The actual implementations of these interfaces change a bit though:

```cs
public class BackgroundTaskQueue : IBackgroundTaskQueue
{
    private readonly ConcurrentQueue<IBackgroundWorkOrder> _workOrders =
        new ConcurrentQueue<IBackgroundWorkOrder>();

    private readonly SemaphoreSlim _signal = new SemaphoreSlim(0);

    public void Queue<TWorkOrder, TWorker>(IBackgroundWorkOrder<TWorkOrder, TWorker> order)
        where TWorker : IBackgroundWorker<TWorkOrder, TWorker>
        where TWorkOrder : IBackgroundWorkOrder<TWorkOrder, TWorker>
    {
        if (order == null)
        {
            throw new ArgumentNullException(nameof(order));
        }

        this._workOrders.Enqueue(order);
        this._signal.Release();
    }

    public async Task<IBackgroundWorkOrder> DequeueAsync(CancellationToken cancellationToken)
    {
        await this._signal.WaitAsync(cancellationToken);
        this._workOrders.TryDequeue(out var workItem);

        return workItem;
    }
}
```

```cs
public class QueuedHostedService : IHostedService
{
    private readonly IServiceProvider _services;

    private readonly CancellationTokenSource _shutdown = new CancellationTokenSource();
    private readonly ILogger _logger;
    private Task _backgroundTask;
    
    public QueuedHostedService(
        IServiceProvider services,
        IBackgroundTaskQueue taskQueue,
        ILoggerFactory loggerFactory)
    {
        this._services = services;
        this.TaskQueue = taskQueue;
        this._logger = loggerFactory.CreateLogger<QueuedHostedService>();
    }

    public IBackgroundTaskQueue TaskQueue { get; }

    public Task StartAsync(CancellationToken cancellationToken)
    {
        this._logger.LogInformation("Queued Hosted Service is starting.");

        this._backgroundTask = Task.Run(this.BackgroundProceessing);

        return Task.CompletedTask;
    }

    private async Task BackgroundProceessing()
    {
        while (!this._shutdown.IsCancellationRequested)
        {
            var workOrder = await this.TaskQueue.DequeueAsync(this._shutdown.Token);

            try
            {
                using (var scope = this._services.CreateScope())
                {
                    var workerType = workOrder
                        .GetType()
                        .GetInterfaces()
                        .First(t => t.IsConstructedGenericType && t.GetGenericTypeDefinition() == typeof(IBackgroundWorkOrder<,>))
                        .GetGenericArguments()
                        .Last();

                    var worker = scope.ServiceProvider
                        .GetRequiredService(workerType);

                    var task = (Task)workerType
                        .GetMethod("DoWork")
                        .Invoke(worker, new object[] { workOrder, this._shutdown.Token });
                    await task;
                }
            }
            catch (Exception ex)
            {
                this._logger.LogError(ex,
                    $"Error occurred executing {nameof(workOrder)}.");
            }
        }
    }

    public Task StopAsync(CancellationToken cancellationToken)
    {
        this._logger.LogInformation("Queued Hosted Service is stopping.");

        this._shutdown.Cancel();

        return Task.WhenAny(
            this._backgroundTask,
            Task.Delay(Timeout.Infinite, cancellationToken));
    }
}
```

In summary:

- easier code-sharing of background work
- proper dependency injection
- allows using external message bus for the queue for many types of background work (excluding `Func<Task>`, not implemented)
- implementation adds complexity, but 

The only real downside I see to this approach is that the worker classes has to be registered with the DI container.

**Comment by rekosko** [comment]
Hi, I'm currently having some serious issues on how to properly register those worker classes... Could you provide an example?

**Comment by vegardlarsen** [comment]
@rekosko You should only need to register the worker class as itself:

```cs
services.AddScoped<BackgroundWorkItem.Worker>();
```

**Comment by rekosko** [comment]
First of all thank you for you reply. I'm using autofac so I've registered it with `containerBuilder.RegisterType<BackgroundWorkItem.Worker>().InstancePerLifetimeScope();` - that worked. I didn't get any exception with `Worker` not being registered but right now some other issue appeared. 

The app is throwing at me `System.ObjectDisposedException` when I'm trying to invoke my service method which tries to query the db. This didn't happen before when I was manually creating scope and getting the correct service implementation.

Any ideas why it would not work?

**Comment by Lissi4kin** [comment]
I've the same problem with autofac:

System.ObjectDisposedException: Cannot access a disposed object. A common cause of this error is disposing a context that was resolved from dependency injection and then later trying to use the same context instance elsewhere in your application. This may occur if you are calling Dispose() on the context, or wrapping the context in a using statement. If you are using dependency injection, you should let the dependency injection container take care of disposing context instances.

**Comment by vegardlarsen** [comment]
@rekosko @Lissi4kin I think your problem is that you are trying to use your `DbContext` from a `QueueBackgroundWorkItem` task?

What is happening is that you are queuing something to happen after the HTTP channel ends. When the HTTP channel ends, the current scope is disposed, and with it goes all the services that were in that scope (e.g. your `DbContext`). So when the time comes to execute your background task, the services you are using are already gone.

If the above explains your situation, this is the exact reason for why I introduced proposed this change; you will need to implement your own `WorkOrder`/`Worker` pair, and get proper dependency injection there. 

So, say you want to send an event that recalculates some expensive property for a user after a change has happened. This is how it would be done:

```cs
public static class UpdateUserReputationBackgroundWork
{
    public static void QueueUpdateUserReputation(this IBackgroundTaskQueue queue, string userId)
    {
        queue.Queue(new Order(userId));
    }

    public class Order : IBackgroundWorkOrder<Order, Worker>
    {
        // This is just the data to pass to the background thread. 
        // No services in here, just a simple data object.
        public Order(string userId)
        {
            this.UserId = userId;
        }

        public string UserId { get; }
    }

    public class Worker : IBackgroundWorker<Order, Worker>
    {
        private readonly DatabaseContext _context;

        public Worker(DatabaseContext context)
        {
            // This is where you put your dependencies, services, etc.
            this._context = context;
        }

        public async Task DoWork(Order order, CancellationToken cancellationToken)
        {
            // This query is not meant to be representative of good code at all, 
            // but just an example of the work to be done, and how we use 
            // dependencies that were injected into the worker.
            var user = await this._context.Users.SingleAsync(u => u.Id == order.UserId);
            user.Reputation = await user.Posts.SumAsync(p => p.Reputation);
            await this._context.SaveChangesAsync();
        }
    }
}
```

You would register this worker as you did the other worker, and then call `backgroundTaskQueue.QueueUpdateUserReputation("vegardlarsen");`

You would implement this kind of pattern for any background task that requires using a service that is registered per scope. Any services that are registered per container lifetime should be fine to use with `QueueBackgroundWorkItem`.

Does that help?

**Comment by Lissi4kin** [comment]
@vegardlarsen yes, it's helped me. Thank you!
